servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/NettyPipelineSslUtils.java

```diff
@@ -54,28 +52,29 @@ public static boolean isSslEnabled(ChannelPipeline pipeline) {
      * @param pipeline the {@link ChannelPipeline} which contains handler containing the {@link SSLSession}.
      * @param sslEvent the event indicating a SSL/TLS handshake completed.
      * @param failureConsumer invoked if a failure is encountered.
+     * @param shouldReport {@code true} if the handshake status should be reported to {@link SecurityHandshakeObserver}.
      * @return The {@link SSLSession} or {@code null} if none can be found.
      */
     @Nullable
     public static SSLSession extractSslSessionAndReport(ChannelPipeline pipeline,
                                                         SslHandshakeCompletionEvent sslEvent,
-                                                        Consumer<Throwable> failureConsumer) {
-        final Channel channel = pipeline.channel();
-        final SecurityHandshakeObserver securityObserver = channel.attr(SECURITY_HANDSHAKE_OBSERVER).get();
+                                                        Consumer<Throwable> failureConsumer,
+                                                        boolean shouldReport) {
+        final SecurityHandshakeObserver observer = shouldReport ? handshakeObserver(pipeline) : null;
         if (sslEvent.isSuccess()) {
             final SslHandler sslHandler = pipeline.get(SslHandler.class);
             if (sslHandler != null) {
                 final SSLSession session = sslHandler.engine().getSession();
-                if (securityObserver != null) {
-                    securityObserver.handshakeComplete(session);
+                if (observer != null) {
+                    observer.handshakeComplete(session);
                 }
```

Comment:
Log at `warn` if `shouldReport` is `true`?

Reply:
The `warn` log is easy to miss. This is not expected to be caught by users, we should provide a guarantee that it never happens. I'm thinking about an assertion, but as we discussed offline we will avoid using `null` values. Therefore, `shouldReport` will go away in a follow-up.

Reply:
Ok what I meant was that if we expect the observer to be present but it is not; is there something we need to do to bubble it up to the user?

Yes good to guarantee but bugs will make it hard to debug. Perhaps a good item to think about when using NOOP observers.